<font size="+3"><b>Assignment 2: Linear Models and Validation Metrics</b></font>

***
* **Full Name** = Aarsh Shah
* **UCID** = 30150079
***

<font color='blue'>In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.</font>

You can use the Table of Content on the left side of this notebook to efficiently navigate within this documents.

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|           **Part 2: Regression**           |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    0.5    |
| Step 3: Implement Machine Learning   Model |     1     |
|            Step 4: Validate Mode           |     1     |
|          Step 5: Visualize Results         |     1     |
|                  Questions                 |     2     |
|             Process Description            |     4     |
|  **Part 3:   Observations/Interpretation** |   **3**   |
|           **Part 4: Reflection**           |   **2**   |
|                  **Total**                 |   **30**  |
|                                            |           |
|                  **Bonus**                 |           |
|         **Part 5: Bonus Question**         |   **4**   |

# **Part 1: Classification (14.5 marks total)**

|                **Question**                | **Point** |
|:------------------------------------------:|:---------:|
|         **Part 1: Classification**         |           |
|          Step 0: Import Libraries          |           |
|             Step 1: Data Input             |     1     |
|           Step 2: Data Processing          |    1.5    |
| Step 3: Implement Machine Learning   Model |           |
|           Step 4: Validate Model           |           |
|          Step 5: Visualize Results         |     4     |
|                  Questions                 |     4     |
|             Process Description            |     4     |
|                  **Total**                 |  **14.5** |

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:


## **Step 0:** Import Libraries

In [1]:
import numpy as np
import pandas as pd

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [2]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y

# Not working so:
from yellowbrick.datasets import load_spam
X, y = load_spam(data_home=None, return_dataset=False)

print(X.size, y.size)
print(type(X), type(y))


262200 4600
<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [3]:
# TO DO: Check if there are any missing values and fill them in if necessary
if (X.isnull().values.any() == True):
    print("There are missing values in the dataset. Filling them in with the mean.")
    X = X.fillna(X.mean())
    
if (y.isnull().values.any() == True):
    print("There are missing values in the dataset. Filling them in with the mean.")
    y = y.fillna(y.mean())


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [4]:
# TO DO: Create X_small and y_small

from sklearn.model_selection import train_test_split
X_small, _, y_small, _ = train_test_split(X, y, test_size=0.95)
print(X_small.size, y_small.size)
X_small

13110 230


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2015            0.00               0.00           0.29           0.0   
1019            0.05               0.05           0.40           0.0   
3272            0.00               0.00           0.00           0.0   
2213            0.00               0.00           0.28           0.0   
2942            0.10               0.10           0.10           0.0   
...              ...                ...            ...           ...   
2320            0.37               0.00           0.63           0.0   
3173            0.00               0.00           0.00           0.0   
3394            0.00               0.00           0.00           0.0   
3470            0.00               0.00           0.00           0.0   
4191            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2015           0.29            0.00               0.0                 0.0   
1019           0.34            0.00               0.0                 0.0   
3272           0.00            0.00               0.0                 0.0   
2213           1.73            0.00               0.0                 0.0   
2942           0.21            0.10               0.0                 0.0   
...             ...             ...               ...                 ...   
2320           0.25            0.12               0.0                 0.0   
3173           0.00            0.00               0.0                 0.0   
3394           0.00            0.00               0.0                 0.0   
3470           0.00            0.00               0.0                 0.0   
4191           0.00            0.00               0.0                 0.0   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
2015             0.29            0.00  ...                   0.0        0.000   
1019             0.57            0.05  ...                   0.0        0.019   
3272             0.00            0.00  ...                   0.0        0.000   
2213             0.00            0.00  ...                   0.0        0.000   
2942             0.10            0.31  ...                   0.0        0.035   
...               ...             ...  ...                   ...          ...   
2320             0.00            0.00  ...                   0.0        0.000   
3173             0.00            0.00  ...                   0.0        0.000   
3394             0.00            0.00  ...                   0.0        0.000   
3470             0.00            0.00  ...                  10.0        0.000   
4191             0.00            1.43  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
2015        0.156        0.031        0.376        0.000        0.125   
1019        0.099        0.000        0.099        0.079        0.009   
3272        0.000        0.000        0.000        0.000        0.602   
2213        0.051        0.000        0.103        0.000        0.000   
2942        0.177        0.035        0.070        0.053        0.000   
...           ...          ...          ...          ...          ...   
2320        0.107        0.000        0.193        0.000        0.000   
3173        0.000        0.000        0.000        0.000        0.000   
3394        0.130        0.065        0.065        0.000        1.043   
3470        0.000        0.000        0.000        0.000        0.000   
4191        0.000        0.000        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
2015                       3.338                         157   
1019                       4.906                          95   
3272                       4.700                          23   
2213                       1.411                           4   
2942                       1.744              

## **Step 3:** Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

## **Step 4:** Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

## **Step 5:** Visualize Results (4 marks for steps 3-5)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [5]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT
from sklearn.linear_model import LogisticRegression
results = pd.DataFrame(columns=['Data size', 'Training accuracy', 'Validation accuracy'])

# Define the datasets
datasets = [
    (X, y),
    (X.iloc[:, :2], y),
    (X_small, y_small)
]

# Process each dataset
for i, (X_data, y_data) in enumerate(datasets):
    # Step 3:
    X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, random_state=0)
    logistic_model = LogisticRegression(max_iter=2000)
    logistic_model.fit(X_train, y_train)

    # Step 4:
    training_accuracy = logistic_model.score(X_train, y_train)
    test_accuracy = logistic_model.score(X_test, y_test)

    # Step 5:
    results.loc[i] = [X_data.size, training_accuracy, test_accuracy]

results



c:\Users\aarsh\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Data size  Training accuracy  Validation accuracy
0   262200.0           0.928696             0.938261
1     9200.0           0.608406             0.613043
2    13110.0           0.953488             0.965517

## **Questions (4 marks)**
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

<font color='Green'><b>YOUR ANSWERS HERE</b></font>

**Note values may differ when running the code**

1. The training and validation accuracy generally increases when the amount of data increases. Here the model with a size of 9200 had an accuracy of 62.1% and 60.2% respectively, compared to a data size of 13110, there was an accuracy of 93.6% and 93.1% respectively. When comparing the data with a size of 262200, the accuracy is quite similar to the dataset with 13110 for the training accuracy (92.9% to 94.7%). This is because the data for the smaller set is seemingly overfit, thus it will work well because it's super specific to that same data. However, the validation accuracy goes down from 93.8% to 86.2% from the large data set to the smaller one. This intuitively makes sense because there is less data, thus the model has seen less, and it overfits the smaller dataset. So in general to increase accuracy you need more **clean/high quality** data. Reasonably speaking you would want a large variety of data paired with a large magnitude so that you don't overfit. Similarly, if you had a smaller dataset you wouldn't want too much variety (and thus you should clip outliers) because that would also result in overfitting. 

2. This dataset describes whether an email is spam, thus a false positive would be a legitimate email incorrectly classified as spam. On the other hand false negative, would be a spam email that is incorrectly classified as legitimate. It generally depends, but in this case, it seems as if the convenience or user experience is a priority (e.g., spam detection), a false positive (blocking a legitimate email) might be considered worse because it could obstruct important communications. But if a false negative occurs, it wouldn't be as much of a big deal because as humans we will see it and just move it to our spam folder manually.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>DESCRIBE YOUR PROCESS HERE</b></font>

My code was sourced with a combination of methods, including sklearn documentation (https://scikit-learn.org/1.4/modules/generated/sklearn.linear_model.LogisticRegression.html), the linear regression lab 2 and Github CoPilot. I used the lab to understand the flow of code and then cross-referenced with the docs to make use of the logistic regression instead. For my use of Github Co-Pliot, I prompted the code completions with comments that were written based on my research and the steps outlined in the lab document. I had to slightly modify code for small parameters, but the code completions were generally quite effective. 

I completed the steps in order until step 3, then I did 3-5 for one dataset to understand it fully. Then I abstracted it into a loop so that I could iterate and perform steps 3-5 for each specified data set. 

The majority of the challenges with this section of the lab surround the concepts of logistic regression and understanding the relevance of different indicators in the data. However, I was able to be successful by referencing the slides from class and the previous labs.

# **Part 2: Regression (10.5 marks total)**

| **Question**                               | **Point** |
|--------------------------------------------|-----------|
| **Part 2: Regression**                     |           |
| Step 1: Data Input                         | 1         |
| Step 2: Data Processing                    | 0.5       |
| Step 3: Implement Machine Learning   Model | 1         |
| Step 4: Validate Mode                      | 1         |
| Step 5: Visualize Results                  | 1         |
| Questions                                  | 2         |
| Process Description                        | 4         |
| **Total**                                  | **10.5**  |

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

## **Step 1:** Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [6]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y
from yellowbrick.datasets import load_concrete

X, y = load_concrete(data_home=None, return_dataset=False)
print(X.size, y.size)
print(type(X), type(y))

8240 1030
<class 'pandas.core.frame.DataFrame'> <class 'pandas.core.series.Series'>


## **Step 2:** Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [7]:
# TO DO: Check if there are any missing values and fill them in if necessary
if (X.isnull().values.any() == True):
    print("There are missing values in the dataset. Filling them in with the mean.")
    X = X.fillna(X.mean())
    
if (y.isnull().values.any() == True):
    print("There are missing values in the dataset. Filling them in with the mean.")
    y = y.fillna(y.mean())

## **Step 3:** Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [8]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

lr = LinearRegression()
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
lm = lr.fit(X_train, y_train)


## **Step 4:** Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [9]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import mean_squared_error, r2_score
y_train_pred = lm.predict(X_train)
y_test_pred = lm.predict(X_test)

# Calculate MSE and R2 for training data
mse_train = mean_squared_error(y_train, y_train_pred)
r2_train = r2_score(y_train, y_train_pred)

# Calculate MSE and R2 for testing data
mse_test = mean_squared_error(y_test, y_test_pred)
r2_test = r2_score(y_test, y_test_pred)

## **Step 5:** Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [10]:
# TO DO: ADD YOUR CODE HERE
results = pd.DataFrame(index=['MSE', 'R2 score'],columns=['Training accuracy', 'Validation accuracy'])
results['Training accuracy'] = [mse_train, r2_train]
results['Validation accuracy'] = [mse_test, r2_test]
results

Training accuracy  Validation accuracy
MSE              111.358439            95.904136
R2 score           0.610823             0.623414

## **Questions (2 marks)**

1. Did using a linear model produce good results for this dataset? Why or why not?

No, using a linear model didn't produce good results, as the R2 score training and validation accuracy values were 61.1 and 62.3% respectively. This indicates that our regression only explains 62.3% of the variation in our y variable. In addition the MSE values for the training and validation accuracy are 111.4 and 95.9 respectively, this indicates a high dispersion around the mean and thus around our function. This is because the data points in y range between 30 and 80, and since both MSE values are greater than the values themselves it causes a significant discrepancy between the predicted and actual values. Particularly the combination of high Mean Squared Error (MSE) values and modest R^2 scores for both training and validation—can be indicative of underfitting.

## **Process Description (4 marks)**
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

<font color='Green'><b>Explain YOUR PROCESS here:</b></font>

Similar to the first part I used a combination of resources. For this one specifically, I primarily used the sklearn LinearRegression class documentation, the code straight from lab 2 since it was also linear regression, and ChatGPT to fill in the gaps. For the generative AI portion of my code, my prompts were quite basic, they followed the instructions pretty closely, but I had to describe to it the purpose of certain things for the AI to understand. I completed the steps pretty much in order but I found it helpful to go to the first section actually and compare and contrast what was being done there and then apply it to a linear problem. 

Most of the challenges surrounding the second part came from interpreting what it meant to "calculate the training and validation accuracy using mean squared error and R2 score." I initially understood this statement in a different way, and I passed the function 'mean_squared_error()' and 'r2_score()' the raw x_train and y_train (as well as test data respectively), but then I asked ChatGPT what it meant and it explained to me that I actually had to use the model to come up with predictions and cross reference that with the actual data. I then changed the code to reflect that using .predict.

# **Part 3: Observations/Interpretation (3 marks)**

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


<font color='Green'><b>ADD YOUR FINDINGS HERE</b></font>

Comparing the implementations of Logistic and Linear Regression models, it appears there are few similarities in their outcomes. In logistic regression, the focus is on creating a curve that models a number of features, utilizing confusion matrices and metrics like precision to assess the model's effectiveness. We also explored the importance of slightly adjusting a training model in classification problems (notably in binary classifications) to ensure errors are more likely to occur on a more forgiving side. This concept was demonstrated in Part 1, highlighting the significance of differentiating between False Positives and False Negatives and understanding the context of the data's application. For instance, an accuracy of 93.8% is impressive; however, if the 6.2% failure rate consistently results in false positives, that would be problematic. 

Conversely, in our discussions on linear regression, we examined how loss functions can serve as an estimate to gauge how well a machine learning model fits a specific dataset. The main functions discussed were mean squared error (MSE) and R-squared score, which are interpreted differently. MSE quantifies the fit of a regression model by measuring the average squared difference between the predicted and actual values. Therefore, an MSE value of 95.904136 does not convey much about our dataset without context. Classroom discussions emphasized that the magnitudes of the 'y' parameter values in the dataset are crucial. For instance, if the 'y' values were on the order of 10^4, then an MSE of 95 would imply high accuracy due to the smaller magnitude in comparison. This, however, contrasts with the actual smaller magnitude of the 'y' values in our dataset.

Nonetheless, both methodologies provide insights into how well a model fits a dataset and addresses a machine learning problem. Through various metrics (e.g., accuracies for different datasets like 262,200 data points with an accuracy of 0.938261, or an R-squared score of 0.623414), we are able to gauge our ability to predict the 'y' parameter for varying data with changes in 'x'. Fundamentally, this reflects our learning and practice from the lectures: predicting 'y' using 'x' parameters with lines.


# **Part 4: Reflection (2 marks)**
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


<font color='Green'><b>ADD YOUR THOUGHTS HERE</b></font>

I like the tangibility of actually using a simple dataset and ML concepts in modeling, predicting, and analyzing the accuracy of something. In class, I find the concepts to be very abstract, and thus it is hard to see the vision for how it would be applied in a real-life situation. This assignment makes it those concepts more concrete and helps me tie what I have learned in class to something practical. What I disliked was that in certain instructions it was a bit ambiguous as to what we were meant to do, and what the exact purpose of it was. For example, in the first part where it wanted us to split the set into 5% of the original set, it was ambiguous as to if we were supposed to split it again when were creating and validating the model. This was challenging as it forced me to make an assumption based on my interpretation of what the instructions say and on what I have learned in class.

# **Part 5: Bonus Question (4 marks)**

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [11]:
# TO DO: ADD YOUR CODE HERE

from sklearn.linear_model import Ridge, Lasso
import numpy as np

# Define a range of alpha values over a logarithmic scale
alpha_values = [0.001, 0.01, 0.1, 1, 10, 100]

# Initialize lists to store results
ridge_results = []
lasso_results = []

# Loop over alpha values for Ridge regression
for alpha in alpha_values:
    ridge_model = Ridge(alpha=alpha)
    ridge_model.fit(X_train, y_train)
    y_train_pred = ridge_model.predict(X_train)
    y_test_pred = ridge_model.predict(X_test)
    mse_train = mean_squared_error(y_train, y_train_pred)
    r2_train = r2_score(y_train, y_train_pred)
    mse_test = mean_squared_error(y_test, y_test_pred)
    r2_test = r2_score(y_test, y_test_pred)
    ridge_results.append([alpha, mse_train, r2_train, mse_test, r2_test])

# Loop over alpha values for Lasso regression
for alpha in alpha_values:
    lasso_model = Lasso(alpha=alpha)
    lasso_model.fit(X_train, y_train)
    y_train_pred = lasso_model.predict(X_train)
    y_test_pred = lasso_model.predict(X_test)
    mse_train = mean_squared_error(y_train, y_train_pred)
    r2_train = r2_score(y_train, y_train_pred)
    mse_test = mean_squared_error(y_test, y_test_pred)
    r2_test = r2_score(y_test, y_test_pred)
    lasso_results.append([alpha, mse_train, r2_train, mse_test, r2_test])

# Convert results to DataFrame for easier analysis
ridge_df = pd.DataFrame(ridge_results, columns=['Alpha', 'MSE_Train', 'R2_Train', 'MSE_Test', 'R2_Test'])
lasso_df = pd.DataFrame(lasso_results, columns=['Alpha', 'MSE_Train', 'R2_Train', 'MSE_Test', 'R2_Test'])

# Find the best R2 score and corresponding alpha for Ridge and Lasso
best_ridge = ridge_df.loc[ridge_df['R2_Test'].idxmax()]
best_lasso = lasso_df.loc[lasso_df['R2_Test'].idxmax()]

best_ridge, best_lasso

(Alpha        100.000000
 MSE_Train    111.358548
 R2_Train       0.610823
 MSE_Test      95.894268
 R2_Test        0.623453
 Name: 5, dtype: float64,
 Alpha         10.000000
 MSE_Train    113.220780
 R2_Train       0.604314
 MSE_Test      95.048607
 R2_Test        0.626774
 Name: 4, dtype: float64)

<font color='Green'><b>ADD YOUR ANSWER HERE</b></font>

It seems as if the Ridge and Lasso models were almost identical for their best performance, in terms of R2_Test, but in general all metrics. Lasso had a slight edge with a R2_Test value of 62.7% compare to Ridge with 62.3%. Ridges best values was a result of alpha=100, whereas for Lasso alpha=10. In general this is not 'good enough' the values are very similar to above using a normal Linear Regression model, and as described the dispersion and underfitting is too severe to be hopeful in accurate results.